### PR TITLE
fix: debounce cookie requests to server

### PR DESCRIPTION
**Pull request by saikumarrs**
## PR Description

Debouncing the cookie requests to the server to avoid repeated calls and cause race conditions while validating the results.
See the ticket for more details.

## Linear task (optional)

https://linear.app/rudderstack/issue/SDK-1960/debounce-server-side-cookie-requests-js-sdk

## Cross Browser Tests

Please confirm you have tested for the following browsers:

- [x] Chrome
- [ ] Firefox
- [ ] IE11

## Sanity Suite

- [x] All sanity suite test cases pass locally

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **New Features**
  - Introduced debounce functionality for server-side cookie settings, enhancing performance.

- **Refactor**
  - Renamed `setServerSideCookie` method to `setServerSideCookies` for clarity and consistency.

- **Bug Fixes**
  - Updated test cases to align with the new method name and added timing-related adjustments for improved accuracy.

### CODE REVIEW

File: packages/analytics-js/src/components/userSessionManager/UserSessionManager.ts [hunk 1]
```diff
@@ -413,13 +423,23 @@ class UserSessionManager implements IUserSessionManager {
...
+                  (cookieName, cookieValue) => {
+                    curStore?.set(cookieName, cookieValue);
+                  },
+                  curStore,
+                );
+              },
+              this,
+              SERVER_SIDE_COOKIES_DEBOUNCE_TIME,
+            );
+          }
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/src/components/userSessionManager/UserSessionManager.ts:440]: Isn't that already done on Ln 426?

File: packages/analytics-js/__tests__/components/userSessionManager/UserSessionManager.test.ts [hunk 2]
```diff
@@ -1501,44 +1501,53 @@ describe('User session manager', () => {
...
+        userSessionManager.setServerSideCookies = jest.fn();
         clientDataStoreCookie.remove = jest.fn();
         userSessionManager.syncValueToStorage('anonymousId', cookieValue);
-        expect(userSessionManager.setServerSideCookie).not.toHaveBeenCalled();
+
+        jest.advanceTimersByTime(1000);
+
+        expect(userSessionManager.setServerSideCookies).not.toHaveBeenCalled();
         expect(clientDataStoreCookie.remove).toHaveBeenCalled();
+        jest.useRealTimers();
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/__tests__/components/userSessionManager/UserSessionManager.test.ts:1545]: It is needed for this test case.

File: packages/analytics-js/__tests__/components/userSessionManager/UserSessionManager.test.ts [hunk 3]
```diff
@@ -1501,44 +1501,53 @@ describe('User session manager', () => {
...
-        expect.any(Function),
-        expect.any(Object),
-      );
+
       setTimeout(() => {
+        expect(setServerSideCookiesSpy).toHaveBeenCalledWith(
+          [{ name: 'rl_anonymous_id', value: 'dummy_anonymousId' }],
+          expect.any(Function),
+          expect.any(Object),
+        );
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/__tests__/components/userSessionManager/UserSessionManager.test.ts:1526]: It feels like an overkill.

File: packages/analytics-js/__tests__/components/userSessionManager/UserSessionManager.test.ts [hunk 4]
```diff
@@ -1696,7 +1705,7 @@ describe('User session manager', () => {
         domain: 'example.com',
         samesite: 'Lax',
       };
-      userSessionManager.setServerSideCookie(
+      userSessionManager.setServerSideCookies(
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/__tests__/components/userSessionManager/UserSessionManager.test.ts:1708]: Invalid suggestion.

File: packages/analytics-js/__tests__/components/userSessionManager/UserSessionManager.test.ts [hunk 5]
```diff
@@ -1586,7 +1595,7 @@ describe('User session manager', () => {
           domain: 'example.com',
           samesite: 'Lax',
         };
-        userSessionManager.setServerSideCookie(
+        userSessionManager.setServerSideCookies(
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/__tests__/components/userSessionManager/UserSessionManager.test.ts:1598]: Invalid suggestion.

**Review by MoumitaM — APPROVED** [review]

**Pull request merged**